__private stuff imported?

Importing a module includes private stuff. I'm not sure that it should?

- `src/a.act`
```python
__foo = {}
```
- `src/test_a.act`
```python
from a import *

__foo = {}

actor main(env):
    print(__foo)
```

So since we do `from a import *` we get `a.__foo` available to us as `__foo` but we already have a `__foo`. This trickles down all the way to C where we hit some error:
```
zig build-lib ActonProject Debug x86_64-linux-gnu.2.27: error: error(compilation): clang failed with stderr: /home/kll/terastream/acton/test/compiler/collide/out/types/test_a.c:79:25: error: use of undeclared identifier 'test_aQ___foo'
```

It's clearly a bug somehow but I don't know which way.
- should `from a import *` actually import __private stuff? I expected it not to
- if it should import private stuff, then I guess the local thing should shadow the import? (but also see #372)

@nordlander ping! :)

--- a/Makefile
+++ b/Makefile
@@ -309,6 +309,9 @@ test-compiler:
 	cd compiler && stack test acton
 	cd compiler && stack test actonc --ta '-p "compiler"'
 
+test-compiler-accept:
+	cd compiler && stack test acton --test-arguments "--golden-start --golden-reset"
+
 test-cross-compile:
 	cd compiler && stack test actonc --ta '-p "cross-compilation"'
 

--- a/compiler/actonc/Main.hs
+++ b/compiler/actonc/Main.hs
@@ -690,7 +690,7 @@ expectedRootStubs paths tasks = do
     return (concat roots)
   where
     mkStub outbase n =
-      if nameToString n == "__test_main"
+      if nameToString n == "test_main"
         then outbase ++ ".test_root.c"
         else outbase ++ ".root.c"
 
@@ -908,7 +908,7 @@ compileFiles gopts opts srcFiles allowPrune = do
         preBinTasks
           | null (C.root opts') = map (\t -> BinTask True (modNameToString (name t)) (A.GName (name t) (A.name "main")) False) rootTasks
           | otherwise        = [binTask]
-        preTestBinTasks = map (\t -> BinTask True (modNameToString (name t)) (A.GName (name t) (A.name "__test_main")) True) rootTasks
+        preTestBinTasks = map (\t -> BinTask True (modNameToString (name t)) (A.GName (name t) (A.name "test_main")) True) rootTasks
     -- Generate build.zig(.zon) for dependencies too, to satisfy Zig builder links.
     let projKeys = Data.Set.fromList (map (tkProj . gtKey) globalTasks)
     forM_ (Data.Set.toList projKeys) $ \p -> do

--- a/compiler/lib/src/Acton/Names.hs
+++ b/compiler/lib/src/Acton/Names.hs
@@ -149,6 +149,14 @@ isHidden _                          = True
 
 notHidden                           = filter (not . isHidden)
 
+isPrivateName                       :: Name -> Bool
+isPrivateName n                     = case nstr n of
+                                        ('_':_) -> True
+                                        _       -> False
+
+isPublicName                        :: Name -> Bool
+isPublicName                        = not . isPrivateName
+
 
 -- Free and bound names ------------
 

--- a/compiler/lib/src/Acton/Syntax.hs
+++ b/compiler/lib/src/Acton/Syntax.hs
@@ -25,7 +25,7 @@ import Control.DeepSeq
 import Prelude hiding((<>))
 
 version :: [Int]
-version = [0,7]
+version = [0,8]
 
 data Module     = Module        { modname::ModName, imps::[Import], mbody::Suite } deriving (Eq,Show,Generic,NFData)
 

--- a/compiler/lib/src/Acton/Types.hs
+++ b/compiler/lib/src/Acton/Types.hs
@@ -76,16 +76,17 @@ reconstruct env0 (Module m i ss)         = do --traceM ("#################### or
         env2                            = define te (setMod m env0)
 
         (teT,ssT)                       = if hasTesting i then (te ++ testEnv, ss1 ++ testStmts env2 (modNameStr m) ss1) else (te, ss1)
-        iface                           = unalias env2 teT
+        iface                           = filterIface (unalias env2 teT)
 
         mrefs                           = moduleRefs1 env0
         env0'                           = convEnvProtos env0
         hasTesting i                    = Import NoLoc [ModuleItem (ModName [name "testing"]) Nothing] `elem` i
+        filterIface                     = filter (isPublicName . fst)
         rmTests (Assign _ [PVar _ n _] _ : ss)
           | nstr n `elem` ["__unit_tests","__simple_sync_tests","__sync_tests","__async_tests","__env_tests"]
                                         = rmTests ss
         rmTests (Decl _ [Actor _ n _ _ _ _ _] : ss)
-          | nstr n == "__test_main"     = rmTests ss
+          | nstr n == "test_main"       = rmTests ss
         rmTests (s : ss)                = s : rmTests ss
         rmTests []                      = []
 
@@ -2188,7 +2189,7 @@ testStmts env m ss                      = ss' ++
         (te, ss')                       = genTestActorWrappers ss
 
 testEnv                                 = [ (name n, NVar (tDict tStr (testing cl))) | (n,cl) <- testDicts ] ++
-                                          [ (name "__test_main", NAct [] posNil (kwdRow (name "env") tEnv kwdNil) [] Nothing) ]
+                                          [ (name "test_main", NAct [] posNil (kwdRow (name "env") tEnv kwdNil) [] Nothing) ]
 
 gname ns n                              = GName (ModName ns) n
 dername a b                             = Derived (name a) (name b)
@@ -2200,7 +2201,7 @@ testing tstr                            = tCon (TC (gname [name "testing"] (name
 mkDict cl as                            = eCall (tApp (eQVar primMkDict) [tStr, testing cl]) [w,Dict NoLoc as]
     where w                             = eCall (eQVar (gname [name "__builtin__"] (dername "Hashable" "str"))) []
 
-testActor                               = sDecl [Actor NoLoc (name "__test_main") []
+testActor                               = sDecl [Actor NoLoc (name "test_main") []
                                                  PosNIL (KwdPar (name "env")  (Just tEnv) Nothing KwdNIL)
                                              [sExpr (eCall (eQVar (gname [name "testing"] (name "test_runner")))
                                                            (map (eVar . name) ["env","__unit_tests","__simple_sync_tests","__sync_tests","__async_tests","__env_tests"]))] Nothing]

--- a/compiler/lib/src/InterfaceFiles.hs
+++ b/compiler/lib/src/InterfaceFiles.hs
@@ -25,7 +25,7 @@
 --   3) ifaceHash    :: ByteString                  -- SHA-256 of public NameInfo (doc-free)
 --                                                 -- augmented with imports' iface hashes
 --   4) imports      :: [(A.ModName, ByteString)]   -- imported module and iface hash used
---   5) roots        :: [A.Name]                    -- root actors (e.g., main or __test_main)
+--   5) roots        :: [A.Name]                    -- root actors (e.g., main or test_main)
 --   6) docstring    :: Maybe String                -- module docstring
 --   7) nameInfo     :: A.NameInfo                  -- type/name environment
 --   8) typedModule  :: A.Module                    -- typed module

--- a/compiler/lib/test/3-types/test_discovery.output
+++ b/compiler/lib/test/3-types/test_discovery.output
@@ -82,5 +82,5 @@ __async_tests: __builtin__.dict[__builtin__.str, testing.AsyncTest] = $mkDict@[_
 
 __env_tests: __builtin__.dict[__builtin__.str, testing.EnvTest] = $mkDict@[__builtin__.str, testing.EnvTest](__builtin__.HashableD_str(), {"_test_EnvTester": testing.EnvTest(_test_EnvTester, "_test_EnvTester", "", "test_discovery")})
 
-actor __test_main (env : __builtin__.Env):
+actor test_main (env : __builtin__.Env):
     testing.test_runner(env, __unit_tests, __simple_sync_tests, __sync_tests, __async_tests, __env_tests)

--- a/compiler/lib/test/ActonSpec.hs
+++ b/compiler/lib/test/ActonSpec.hs
@@ -38,6 +38,7 @@ import System.FilePath ((</>), joinPath, takeFileName, takeBaseName, takeDirecto
 import System.Directory (getCurrentDirectory, setCurrentDirectory)
 import Control.Monad (forM_, when, foldM)
 import qualified Control.Exception as E
+import Control.DeepSeq (rnf)
 import Utils (SrcLoc(..), loc, prstr)
 import qualified Acton.BuildSpec as BuildSpec
 import qualified Data.ByteString.Lazy as BL
@@ -448,6 +449,45 @@ main = do
       testTypes env0 ["test_discovery"]
 
       testAttributesInitialization env0
+
+    describe "Import Semantics" $ do
+      it "omits private names from the public interface" $ do
+        (envA, parsedA) <- parseAct env0 "import_private_a"
+        kcheckedA <- liftIO $ Acton.Kinds.check envA parsedA
+        (nmodA, _, _, _) <- liftIO $ Acton.Types.reconstruct envA kcheckedA
+        let S.NModule tenvA mdocA = nmodA
+            env1 = Acton.Env.addMod (S.modname parsedA) tenvA mdocA env0
+
+        (envB, _) <- parseAct env1 "import_private"
+        let namesB = Acton.Env.names envB
+
+        case lookup (S.name "__foo") namesB of
+          Nothing -> pure ()
+          Just _ -> expectationFailure "from import * should skip __foo"
+
+        case lookup (S.name "public_value") namesB of
+          Just (S.NAlias _) -> pure ()
+          _ -> expectationFailure "from import * should include public_value"
+
+      it "blocks qualified access to private names" $ do
+        (envA, parsedA) <- parseAct env0 "import_private_a"
+        kcheckedA <- liftIO $ Acton.Kinds.check envA parsedA
+        (nmodA, _, _, _) <- liftIO $ Acton.Types.reconstruct envA kcheckedA
+        let S.NModule tenvA mdocA = nmodA
+            env1 = Acton.Env.addMod (S.modname parsedA) tenvA mdocA env0
+
+        (envB, parsedB) <- parseAct env1 "import_private_qualified"
+        kcheckedB <- liftIO $ Acton.Kinds.check envB parsedB
+        result <- liftIO $ (E.try (do
+          (_, tcheckedB, _, _) <- Acton.Types.reconstruct envB kcheckedB
+          _ <- E.evaluate (rnf tcheckedB)
+          pure ()
+          ) :: IO (Either CompilationError ()))
+
+        case result of
+          Left NoItem{} -> pure ()
+          Left err -> expectationFailure $ "Expected NoItem error, got " ++ show err
+          Right _ -> expectationFailure "Expected type check failure for private name access"
 
     describe "Pass 4: Normalizer" $ do
       testNorm env0 ["deact"]

--- a/compiler/lib/test/src/import_private.act
+++ b/compiler/lib/test/src/import_private.act
@@ -1,0 +1,6 @@
+from import_private_a import *
+
+actor main(env):
+    print(__foo)
+    print(public_value)
+    env.exit(0)

--- a/compiler/lib/test/src/import_private_a.act
+++ b/compiler/lib/test/src/import_private_a.act
@@ -1,0 +1,2 @@
+__foo: int = 1
+public_value: int = 2

--- a/compiler/lib/test/src/import_private_qualified.act
+++ b/compiler/lib/test/src/import_private_qualified.act
@@ -1,0 +1,5 @@
+import import_private_a
+
+actor main(env):
+    print(import_private_a.__foo)
+    env.exit(0)

